src/main/java/net/fabricmc/loom/task/MigrateMappingsTask.java

```diff
@@ -61,123 +68,130 @@ public void doTask() throws Throwable {
+		if (outputDir == null) outputDir = "remappedSrc";
+		String targetMappingsVersion = (String) properties.get("targetMappings");
 
-		if (!properties.containsKey("inputDir") || !properties.containsKey("outputDir")) {
-			throw new RuntimeException("Must specify input and output dir!");
+		if (targetMappingsVersion == null) {
+			throw new IllegalArgumentException("You must specify a new mappings version with -PtargetMappings.");
 		}
 
-		File inputDir = new File((String) properties.get("inputDir"));
-		File outputDir = new File((String) properties.get("outputDir"));
+		Path inputDirPath = Paths.get(System.getProperty("user.dir"), inputDir);
+		Path outputDirPath = Paths.get(System.getProperty("user.dir"), outputDir);
 
-		if (!inputDir.exists() || !inputDir.isDirectory()) {
-			throw new RuntimeException("Could not find input directory: " + inputDir);
+		if (!Files.exists(inputDirPath) || !Files.isDirectory(inputDirPath)) {
+			throw new IllegalArgumentException("Could not find input directory: " + inputDirPath.toAbsolutePath());
 		}
 
-		if (!outputDir.exists()) {
-			if (!outputDir.mkdirs()) {
-				throw new RuntimeException("Could not create output directory:" + outputDir);
-			}
-		}
+		Files.createDirectories(outputDirPath);
 
-		Mappings sourceMappings = extension.getMappingsProvider().getMappings();
-		Mappings targetMappings;
+		MappingsProvider mappingsProvider = extension.getMappingsProvider();
 
-		try (FileInputStream stream = new FileInputStream(mappingsFile)) {
-			targetMappings = net.fabricmc.mappings.MappingsProvider.readTinyMappings(stream, false);
+		try {
+			TinyTree currentMappings = mappingsProvider.getMappings();
+			TinyTree targetMappings = mappingsProvider.getMappingsOfVersion(project, targetMappingsVersion);
+			migrateMappings(project, extension.getMinecraftMappedProvider(), inputDirPath, outputDirPath, currentMappings, targetMappings, extension);
+			project.getLogger().lifecycle(":remapped project written to " + outputDirPath.toAbsolutePath());
+		} catch (IOException e) {
+			throw new IllegalArgumentException("Could not find mappings for version " + targetMappingsVersion, e);
 		}
+	}
 
+	private static void migrateMappings(Project project, MinecraftMappedProvider minecraftMappedProvider,
+										Path inputDir, Path outputDir, TinyTree currentMappings, TinyTree targetMappings, LoomGradleExtension extension
+	) throws IOException {
 		project.getLogger().lifecycle(":joining mappings");
-		MappingSet mappingSet = new MappingsJoiner(sourceMappings, targetMappings, "intermediary", "named").read();
+		MappingSet mappingSet = new MappingsJoiner(currentMappings, targetMappings,
+						"intermediary", "named").read();
 
 		project.getLogger().lifecycle(":remapping");
-		Mercury mercury = new Mercury();
+		Mercury mercury = SourceRemapper.createMercuryWithClassPath(project, false);
 
-		for (File file : project.getConfigurations().getByName(Constants.MINECRAFT_DEPENDENCIES).getFiles()) {
```

Comment:
@modmuss50 I'm not sure why I removed this... Does this seem ok?

Reply:
Oh it's embedded in `SourceRemapper.createMercuryWithClassPath(project, false);`, right